core/test/RectGrid_test.cpp

```diff
@@ -104,47 +105,95 @@ TEST_CASE("Write and read a ModelState-based RectGrid restart file")
     metadata.affixCoordinates(state);
 
 #ifdef USE_MPI
+// Create subcommunicator with only first rank
     metadata.setMpiMetadata(test_comm);
-    metadata.globalExtentX = nx;
-    metadata.globalExtentY = ny;
-    metadata.localCornerX = 0;
-    metadata.localCornerY = 0;
-    metadata.localExtentX = nx;
-    metadata.localExtentY = ny;
-#endif
+    int colour = MPI_UNDEFINED, key = 0;
+    MPI_Comm rank0Comm;
+
+    if(metadata.mpiMyRank ==0) {
+        colour = 0;
+    }
+    MPI_Comm_split(test_comm, colour, key, &rank0Comm);
+
+// Write reference file serially on first MPI rank
+    if (metadata.mpiMyRank == 0) {
+        metadata.setMpiMetadata(rank0Comm);
+        metadata.globalExtentX = nx;
+        metadata.globalExtentY = ny;
+        metadata.localCornerX = 0;
+        metadata.localCornerY = 0;
+        metadata.localExtentX = nx;
+        metadata.localExtentY = ny;
+        grid.dumpModelState(state, metadata, filename);
+        MPI_Comm_free(&rank0Comm);
+    }
+#else
     grid.dumpModelState(state, metadata, filename);
+#endif
+
 
     ModelArray::setDimensions(ModelArray::Type::H, { 1, 1 });
+    ModelArray::setDimensions(ModelArray::Type::Z, { 1, 1, 1 });
     REQUIRE(ModelArray::dimensions(ModelArray::Type::H)[0] == 1);
     RectangularGrid gridIn;
-    size_t targetX = 3;
-    size_t targetY = 7;
+    size_t targetX = 1;
+    size_t targetY = 2;
 
     gridIn.setIO(new RectGridIO(grid));
 #ifdef USE_MPI
-    ModelState ms = gridIn.getModelState(filename, partition_filename, metadata);
+    ModelMetadata metadataIn;
+    metadataIn.setTime(TimePoint(date_string));
+    metadataIn.setMpiMetadata(test_comm);
+    ModelState ms = gridIn.getModelState(filename, partition_filename, metadataIn);
 #else
     ModelState ms = gridIn.getModelState(filename);
 #endif
 
+#ifdef USE_MPI
+    REQUIRE(ModelArray::dimensions(ModelArray::Type::H)[0] == metadataIn.localExtentX);
+    REQUIRE(ModelArray::dimensions(ModelArray::Type::H)[1] == metadataIn.localExtentY);
+    REQUIRE(ModelArray::dimensions(ModelArray::Type::Z)[0] == metadataIn.localExtentX);
+    REQUIRE(ModelArray::dimensions(ModelArray::Type::Z)[1] == metadataIn.localExtentY);
+#else
     REQUIRE(ModelArray::dimensions(ModelArray::Type::H)[0] == nx);
     REQUIRE(ModelArray::dimensions(ModelArray::Type::H)[1] == ny);
+    REQUIRE(ModelArray::dimensions(ModelArray::Type::Z)[0] == nx);
+    REQUIRE(ModelArray::dimensions(ModelArray::Type::Z)[1] == ny);
+#endif
     REQUIRE(ms.data.at("hice")(targetX, targetY) != 0);
     REQUIRE(ms.data.at("hice")(targetX, targetY) > 1);
     REQUIRE(ms.data.at("hice")(targetX, targetY) < 2);
-    REQUIRE(ms.data.at("hice")(targetX, targetY) == 1.0703);
+#ifdef USE_MPI
+    REQUIRE(ms.data.at("hice")(targetX, targetY) == 1.0201 + metadataIn.localCornerY * 0.01 + metadataIn.localCornerX * 0.0001);
```

Comment:
Where do these numbers come from?

Reply:
Good catch. They should be `xFactor` and `yFactor. I will fix it.